### PR TITLE
Create Task Service

### PR DESCRIPTION
**Pull request by Deolamma**
In this PR the following were implemented:

- Renamed the parent folder of taskService and axiosConfig from api to services
- Added filtering by queryParams to getAllTasks
- Updated the Task Service methods to reuse getTask method where necessary fostering DRY principle
- Implemented the deleteNotes method and added a test for it in the file below. **NOTE:** to use this file rename the extension from .txt to .js
[testEndpoints.txt](https://github.com/user-attachments/files/16398991/testEndpoints.txt)
- Updated the ID's of Tasks, subtasks, and notes to UUID's in the data.json file
- Updated the Subtasks to contain `updatedAt` field properties 
- Created `UtilityFunctions.js` to handle filtering of params and server errors
- Updated taskService methods to perform CRUD ops based on the new db structure

**Video of service methods being tested:**
https://www.awesomescreenshot.com/video/29943982?key=76e09a9072a5f25aaf6d762ea0212a41

**Checkout the description in this PR for more info:** https://github.com/adebisit/simple-task-mgmt-app/pull/9

### CODE REVIEW

File: src/services/taskService.js [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+/**
+ * getAllTasks - Asynchronous function that gets all tasks in the db
+ * @param {object} queryParams: An object that contains the query parameters to retrieve tasks with
+ *                              or defaults to an empty object
+ * @returns {Promise<Array>} All tasks created
+ */
+export const getAllTasks = async (queryParams = {}) => {
+  try {
+    //Building the queryString...
+    const queryString = new URLSearchParams(queryParams).toString();
```
Comment (adebisit) [inline src/services/taskService.js]: Since we might not be able to filter by query using the "like" keyword and JSON-server doesnt filter date range; you might want to implement logic to manually filter those (if its included in the queryParams) after getting the result from JSON server.
Also while testing I realized if you included these two queries I mention; it makes the url invalid and returns an empty array.

So essentially your function should
1. Remove the keyword and date range parameters from the query Param first
2. Perform your request like you already are
3. Filter the keywork and date range parameters manually from the response.
4. return said response.

Do you understand?

File: src/services/taskService.js [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+    //check status of response
+    if (res.status === 500) {
+      throw new Error("Server error");
+    }
+    //retrieve task from response
+    const task = res.data;
+    if (!task) {
+      throw new Error("Task not found");
+    }
+    return task;
```
Comment (adebisit) [inline src/services/taskService.js:55]: Good Error handling.

File: src/services/taskService.js [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+  }
+};
+
+/**
+ * getSubtask - Async function that retrieves a single subtask
+ * @param {String} taskId ID of task containing the subtask to be retrieved
+ * @param {String} subtaskId ID of subtask to be retrieved from subtasks array
+ * @returns {object} subtask to be retrieved
+ */
+export const getSubtask = async (taskId, subtaskId) => {
```
Comment (adebisit) [inline src/services/taskService.js]: Is there any situation where we'd need to get a specific subtask?

The Figma design has a section that shows all subtasks at once. Same applies with notes. Unless we have a specific subtask detail page; this might not be needed.

Do you understand?

**Comment by adebisit** [comment]
I left some comments on Whatsapp on something I think would change the way we approach CRUD of subtasks and notes. If you test it out and it is possible; (you and @Haydox01) please create a PR of the new branches first and I'd comment and merge before I merge this one.

**Review by adebisit — CHANGES_REQUESTED** [review]

File: src/services/taskService.js (hunk 3 above)
Comment (Deolamma) [reply src/services/taskService.js]: Not at the moment, I will take them out.

File: src/services/taskService.js (hunk 1 above)
Comment (Deolamma) [reply src/services/taskService.js]: Got it.

**Review by adebisit — APPROVED** [review]
